URAP-charter/web_scraping
# Hotfix george test hang issue

The purpose of this branch was to identify and examine the issue of "hanging" on certain URLs from the achievementfirst.org domain. The issue that I found was: These URLs were direct file download links, even though they didn't have the ".pdf" at the end of their URLs. 

Using the application "Postman", I was able to confirm this theory:
![image](https://user-images.githubusercontent.com/65363191/110510475-1172b500-80b8-11eb-8b0c-462dccaf84fd.png)
Where you can see that the "Content-Type" is "application/pdf".

This branch therefore attempts to resolve the issue via a try/catch clause in the "get files" and "get images" methods in scrapy_vanilla.py. NOTE: Try/Catch should NOT be used for flow-control. This is just to show what the issue is and put a band-aid on it. To prevent the files from being parsed as regular text (which results in jumbled up gibberish), I also attempted to catch the correct content types to process the links as we do normal PDF documents. However, I encountered another issue which will likely be another ticket: these files are likely some kind of download stream, so they are not processing correctly and will likely need some extra work to get them neatly into our DB.

The purpose of this pull request is to resolve the hanging part of the issue though, which should not be a problem.

To test this:
1. Run MongoDB somewhere accessible (I recommend `docker pull mongo && docker run -p 27017:27017 --name mongodb mongo`, but do it however you like best) and connect to it via settings.py (ensure that the MONGO_URI is the URI where your Mongo lives -- localhost:27017 if you ran it how I suggested). 
2. Start your venv as explained in the README
3. Get the test_50.tsv file from the ticket: https://trello.com/c/4rOpRLLi/131-debug-scrapy-hanging-issue-1-identify-issue
4. Run scrapy: `scrapy crawl schoolspider...` (as shown in the README -- the school_list parameter should be a path to your test_50.tsv file)
5. This will take a while to run, so be patient!
6. At some point, check your run and you'll see a lot of "achievementfirst.org/download/[some number]" urls being parsed. These URLs were the ones that were hanging. Note that it is slowly trying to parse them (it will likely run into some issues due to the file streaming that may be used at the URLs, but it will move on and try the next URL)
7. At this point, you can wait for it to run to completion or stop it to save memory and sanity
8. Check your MongoDB -- the outputItems collection of the schoolSpider db -- and notice that there's a LOT more information there than other runs (for example when running my other ticket to test concurrent scrapy runs, only like 2 items are written in the first 1k URLs)

At this point, we can decide whether to merge this branch as-is, or if the streaming PDF issue should be resolved first
